src/main.cpp

```diff
@@ -3661,13 +3661,34 @@ bool CheckColdStakeFreeOutput(const CTransaction& tx, const int nHeight)
         if (lastOut.nValue == 3 * COIN)
             return true;
 
+        // This could be a budget block.
+        if (Params().IsRegTestNet())
+            return false;
+
+        // if mnsync is incomplete, we cannot verify if this is a budget block.
+        // so we check that the staker is not transferring value to the free output
+        if (!masternodeSync.IsSynced()) {
+            // First try finding the previous transaction in database
+            CTransaction txPrev; uint256 hashBlock;
+            if (!GetTransaction(tx.vin[0].prevout.hash, txPrev, hashBlock, true))
+                return error("%s : read txPrev failed: %s",  __func__, tx.vin[0].prevout.hash.GetHex());
+            CAmount amtIn = txPrev.vout[tx.vin[0].prevout.n].nValue + GetBlockValue(nHeight - 1);
+            CAmount amtOut = 0;
+            for (unsigned int i = 1; i < outs-1; i++) amtOut += tx.vout[i].nValue;
+            if (amtOut < amtIn)
+                return error("%s: non-free outputs value %d less than required %d", __func__, amtOut, amtIn);
```

Comment:
Shouldn't we checking that `amtOut == amtIn`?

Reply:
If `amtOut > amtIn` the transaction is not valid anyway and the block overmints. But yeah, it doesn't hurt to be more strict directly from here. Will update this.

Reply:
pushed